### ISSUE TITLE
Dump state into new genesis file

### ISSUE THREAD

**Issue by atenjin** [paritytech/substrate#2116]
I think substrate may need a checkpoint feature.
it could be used in follow situation
* sync block from some checkpoint, not from genesis
* could export all key/value from some checkpoint(or could export all node's key in trie tree for checkpoint state root), and then, could use this exported key/value or other data be a new genesis. Using this exported data, we could restart a new chain to recover from some crisis problem(though, it's very ugly, **but it maybe a final plan**, just like what cosmos recently do, cosmos export data from 425000, change code, start a new chain. [link](https://hubble.figment.network/chains/cosmoshub-1/governance/proposals/2) )
* maybe others...

**Comment by xlc** [comment]
Another similar feature I want is able to generate genesis file from current state. This means I can configure the testing environment once (deposit funds, deploy smart contract, doing configuration that is not possible with GenesisConfig, etc) and make a genesis file from it, and reuse it.

**Comment by tomaka** [comment]
Note that it is already possible to export and import blocks out of/into the database, if that's what this issue is about.

**Comment by apopiak** [comment]
The `substrate export-state` subcommand introduced in https://github.com/paritytech/substrate/pull/5842 should solve this, right?

**Comment by bkchr** [comment]
Not entirely. We still don't have checkpoints.

**Comment by atenjin** [comment]
> The `substrate export-state` subcommand introduced in #5842 should solve this, right?

not the full feature. for now, the export-state just can start the new chain from 0 height. but in fact, the best way is to start the new chain with the same height.

I think it needs a lot of modification for substrate backend part, and related to some consensus part.

**Comment by takahser** [comment]
According to [Kusama Network](https://www.youtube.com/c/KusamaNetwork)'s YouTube video _[Part 1: How to Register a Parachain on Polkadot and Kusama](https://www.youtube.com/watch?v=BaWcQxKRFmM)_ it should be possible to generate the genesis file and the runtime wasm file by running the following command:

<img width="925" alt="image" src="https://user-images.githubusercontent.com/5393704/192701100-c95ab87e-08ab-4dbc-b13c-9a8b8b6578bc.png">

I tried that using the [node-template](https://github.com/substrate-developer-hub/substrate-node-template) however, I got an error:

```bash
$ ./node-template export-genesis-state --parachain-id 21; ./node-template export-genesis-wasm > wasmfile
error: The subcommand 'export-genesis-state' wasn't recognized

	Did you mean 'export-blocks' or 'export-state'?
```

Also, when I try to run the suggested command, the `parachain-id` arg is not recognized:

```bash
 % ./node-template export-state --parachain-id 2225;
error: Found argument '--parachain-id' which wasn't expected, or isn't valid in this context
```

So my questions are:

1. What's the correct command for exporting the genesis state?
2. Also, what's the correct command for exporting the runtime wasm, since the `export-genesis-wasm` command is not recognized at all?

   ```bash
   % ./node-template export-genesis-wasm > wasmfile
   error: The subcommand 'export-genesis-wasm' wasn't recognized
   ```
